src/Shuttle.h

```diff
@@ -56,41 +58,53 @@ class CommandParameters;
 \brief Shuttle that deals with parameters.  This is a base class with lots of
 virtual functions that do nothing by default.
 Unrelated to class Shuttle.
+
+@tparam Const if true, then visited settings are not modifiable.
 ********************************************************************************/
-class AUDACITY_DLL_API SettingsVisitor /* not final */
+template<bool Const>
+class SettingsVisitorBase /* not final */
 {
 public:
+   template<typename T> using Ref = std::conditional_t<Const, const T&, T&>;
+   template<typename T> using Arg = std::conditional_t<Const, T, T&>;
+
```

Comment:
I am a bit confused by the purpouse of Arg<>. 
I would have expected it to be defined like `std::conditional_t<Const, const T, T>;` - but as it is now, it switches passing an arg either by copy or by nonconst-reference; I find it strange that the bool Const in one case (Ref) effectively switches constness, and in the other (Arg) switches between copy/reference.

Reply:
I explained why the second alternative must be `T&` to allow non-const visitors to mutate the effect they are visiting.

The difference between Ref and Arg for const visitors is just about the efficiency of pass by value versus reference.

Making the first alternative of arg `Const T` doesn't really change the meaning of the function declarations that use it.  `const` qualifying a by-value argument in a declaration really makes no difference to how the calling code compiles.  It makes a difference only in the function definition, to treat the argument as an immutable local variable.

Reply:
Alright, I see it now in this way: Ref and Arg are both ways to statically switch between passing an argument with a no-change guarantee, or not. 

Arg differs by Ref in this way: the no-change-guarantee is implemented with pass-by-value. Use one or the other, depending on what is being passed.